### PR TITLE
add globals to asm-to-wasm

### PR DESCRIPTION
**Pull request by aseemgarg89**
*globals can't be initialized yet

### CODE REVIEW

File: src/wasm/asm-wasm-builder.cc [hunk 1]
```diff
@@ -355,21 +376,22 @@ class AsmWasmBuilderImpl : public AstVisitor {
...
-        return;
+    if (in_function_) {
+      BinaryOperation* value_op = expr->value()->AsBinaryOperation();
+      if (value_op != NULL && MatchBinaryOperation(value_op) == kAsIs) {
+        VariableProxy* target_var = expr->target()->AsVariableProxy();
+        VariableProxy* effective_value_var = GetLeft(value_op)->AsVariableProxy();
+        if (target_var != NULL && effective_value_var != NULL &&
+            target_var->var() == effective_value_var->var()) {
+          block_size_--;
+          return;
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: How about just emitting a nop and doing away with all this block_size logic?

File: src/wasm/asm-wasm-builder.cc [hunk 2]
```diff
@@ -288,14 +294,29 @@ class AsmWasmBuilderImpl : public AstVisitor {
...
+          } else {
+            current_function_builder_->AppendCode(kExprGetLocal, false);
+          }
         }
         LocalType var_type = TypeOf(expr);
         DCHECK(var_type != kAstStmt);
-        AccessTemp(LookupOrInsertLocal(var, var_type));
+        if (var->IsContextSlot()) {
+          uint32_t index = LookupOrInsertGlobal(var, var_type);
+          std::vector<uint8_t> index_vec = UnsignedLEB128From(index);
```
Comment (titzer) [inline src/wasm/asm-wasm-builder.cc]: This is kind of awkward and relatively inefficient. How about an AddLEB128() method and hiding the encoding inside there?

File: src/wasm/asm-wasm-builder.cc (hunk 1 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: I am not sure if that's a good idea. The whole reason to even do this is to reduce the size of the binary. Right now we are removing 3 bytes for every param (I think that is the only significant use case where this optimization gives any benefit). If we move to NOP, we will be removing 2 instead. However, with all the other possible inefficiencies in there, I am not sure how much this optimization is worth. Will let you make a final call

File: src/wasm/asm-wasm-builder.cc (hunk 2 above)
Comment (aseemgarg89) [inline src/wasm/asm-wasm-builder.cc]: done. refactored and reused AccessTemp

File: src/wasm/asm-wasm-builder.cc (hunk 1 above)
Comment (flagxor) [inline src/wasm/asm-wasm-builder.cc]: Agreed the block_size logic is hard to follow and might be premature optimization, but let's separate switching that round to another CL (as I'd like to build on this one).

**Comment by flagxor** [comment]
lgtm with one nit and maybe add a TODO to either make the block_size dance more clear or suffer an extra NOP.

**Comment by flagxor** [comment]
lgtm
